Xenomorph Mode

Super3222 opened the pull request:
## Xenomorph Mode

This game mode is designed to be a strategic battle royal between the xenomorphs and the crew. It starts off with the Queen spawning in random maintenance areas guarded by several random spawning xenomorphs. The schematics are set so within five or so minutes a random announcement will alert the crew of the xenomorph presence.

### Checklist

- [x] Gamemode
- [x] Balancing Out the Xenomorphs
- [x] Design Doc Treading
- [x] Balancing
- [x] Fix Horrific Merge Conflicts.
- [x] Proper Predator Objectives
- [x] Alternatives to Xenomorph Objective/Goal

## Design Doc

https://docs.google.com/document/d/14LjKCb1yBz1cE24vr-cSYjK4Mg6Q7ZjQWWcx-Yv9mUI/edit?usp=sharing

## Pictures / Footage of Gameplay

![](http://image.prntscr.com/image/0c71bc2d7e084996ad7ba5d5fa661d64.png)

![](http://image.prntscr.com/image/f351305b710b44779a354fd8f8e9156b.png)

![](http://image.prntscr.com/image/d22b197a249e4c348a733281879e6a1d.png)

![](http://image.prntscr.com/image/bc37063a6db443fd939edc4a07c083d6.png)

![](http://image.prntscr.com/image/1a30eee53dea4218aff1913876231881.png)

https://i.gyazo.com/3bfe23d6312908a03ac97c244a17df87.mp4

#### Changelog

:cl:
experimental: Xenomorph Mode is now on Yogstation!
rscadd: The Aliens that you know have been massively tweaked, and updated with new features. The list is so incredibly long that it's better to find out for yourself in game. To give you an idea we have new xenomorphs, new abilities, new skills, new organs, new stats, new colors, and a whole new mode for them to thrive in.
rscadd: Alien Queen and her hive spawn somewhere in maintenance with their one team objective to complete.
rscadd: Currently there is two distinct team objectives for Xenomorph Mode. The first one is the 'Conquer' objective where the goal is to spread a certain amount of weeds across the station to dominate it.
rscadd: The second one is the 'Infestation' objective where the goal is to infect a certain percentage of the crew.
rscadd: Predators will be making their very first official landing on Yogstation. You will find them dropping into the game when the xenomorphs have accomplished 45% of their objective. These Predators come packed with their own special weapons and goals to accomplish while the crew is on board. For better or worse.
/:cl:

Kmc2000 commented:
HOOOOO
👌👀👌👀👌👀👌👀👌👀 good shit go౦ԁ sHit👌 thats ✔ some good👌👌shit right👌👌there👌👌👌 right✔there ✔✔if i do ƽaү so my self 💯 i say so 💯 thats what im talking about right there right there (chorus: ʳᶦᵍʰᵗ ᵗʰᵉʳᵉ) mMMMMᎷМ💯 👌👌 👌НO0ОଠOOOOOОଠଠOoooᵒᵒᵒᵒᵒᵒᵒᵒᵒ👌 👌👌 👌 💯 👌 👀 👀 👀 👌👌Good shit

ShadowDeath6 commented:
half competent xenos will steamroll since it's a waste of time to clear weeds
the xeno on the bottom looks really weird, like it's black and white

matskuman5 commented:
cool

Jc1469 commented:
balance stuff I went over with Adam on  :thinking:  because adam knows how to balance stuff

AdamElTablawy commented:
none of this is stuff i went over tho

Time-Green commented:
I am not sure about the stingers, it's non-cannon and triggering me.

Why is the xeno on the picture covered in semen?

You don't need to do a trilobite and deacon, as the trilobite came from the mutagen.

The aliens were space proof, just not to instant temperature changes.

PREDALIENS

CORGALIENS

Remove the ability to throw facehuggers and have aliens kidnap people to the nest instead?

Wall hives?

Apart from the some stuff being non-cannon, amazing work super 👍

ShadowDeath6 commented:
>canon 

Uhh sorry these aliens have no actual relation to any creative or artistic creation, these were taken from tgstation's alien mini-antag which were created from scratch with no influence by any outside forces.

Lightning38088 commented:
Whats the bite do? Will it like insta kill someone or what? Also I feel like they should just spawn in maint lIke usual since I don't see where else they'd spawn.

Kmc2000 commented:
@Lightning38088

Iirc, it gibs the target's head

AdamElTablawy commented:
already going over stuff time

Mrcreepssss commented:
i reccomend to look at colonial marines (probs already did) to see what you like, perhaps even port some of the stuff. further i do reccomend to change objective possibilities to have yee many morphs in the hive, kill objectives like tator has, escape with this many morphs, hijack shuttle all that sort of stuff instead. or perhaps even a kill everyone objective.

Slinky5890 commented:
> kill objectives like tator has

I agree with your ideas apart from this. Too easy and quick

X-TheDark commented on code/__HELPERS/icons.dm:
```diff
@@ -897,6 +897,21 @@ var/global/list/friendly_animal_types = list()
+	flat_icon.BecomeAlphaMask()
+	var/icon/blank_icon = new/icon('icons/effects/effects.dmi', "invsthermal_base")
+	blank_icon.AddAlphaMask(flat_icon)
+
```
Missing `return blank_icon`

Super3222 replied:
Whoops!

Mrcreepssss commented:
@Slinky5890 
i guess that depends on what role and their competence, maybe have them target a ling? or have multiple kill objectives.

(also how do you quote text like that?)

Mrcreepssss commented:
also, with hijack we're gonna have to find a solution for if no-one manages to call the shuttle and no-one will be able to call it we could hardly count on the end of shift shuttle?
maybe some kind of distress beacon mechanic that can be activated from multiple areas on the station that'd increase likelyhood of certain high-tech-ish midround antags and brings a chance that ERT gets sent if online for yee long, it should also make an announcement that tells it's activated. infact, this would be a cool mechanic in any scenario. though it'd need to be balanced for all sorts of antags.

Daman453 commented:
Explain how the sound system works for humans?

Super3222 commented:
> Explain how the sound system works for humans?

I tried explaining it better in the OP. Hope that helps.

------
As for progress, I cleaned all of the merge conflicts last week. Right now I'm working on implementing balance changes and features in my own way. My next commit will hopefully make the mode ready for live testing (before or after code review).

ShadowDeath6 commented:
;_;7

Super3222 commented:
# FIXED 99% OF THE CONFLICT ERRORS
------

Numerical Changes:
Pounce Stun: 5 -> 3
Facehugger throw range: 5 -> 2

Tested Changes (in no particular order):
- Facehuggers can no longer tear off certain masks. These masks have to be completely covering your face, like a generic gas mask (NOT the breath mask you get in your boxes) and a syndicate gas mask.
- Dead hosts no longer breed live larvas. Instead they shit out dried, unplayable ones and get their bodies gibbed.
- Hunter speed is nerfed some more. It works on intervals, where there's a 20% chance that it will transfer faster every time it enters a turf. Confusing, but it works. If you don't understand it don't try. Just know that it'll make hunters slower. Hunters are fairly fast now.
- Chestbursting now deletes your brain (yes, this is intended so that you can't get back into the round).
- Ports over @X-TheDark  revised heart sensor code (Predator). Thanks goes out to him.
- Combines the senitel and drone classes, and calls them xenomorph workers. These workers goal is pretty similar to both functions of the senitel and drone classes. They operate to support the inner hive rather than going out and fighting.
- Exiting cloak mode as a Predator makes a really obvious noise. Hopefully will help xenomorphs and humans out abroad.
- Added game logging to objectives (global and direct), and hive balancing.
- Workers and the Queen can pick up items in the game.
- Hunters cannot tackle stunned people.
- Predators now spawn with objectives. There are 4 different types. 1) Trophy - the Predators goal is to capture the head of a specific command member. Two predators cannot chase the same person. 2) Queen - Assassinate the Queen. 3) Capture X amount of hunter tails. 4) Capture X amount of worker tails. Predators will have a 50/50 shot at getting either the hunter/worker objective. 75/25 for trophy/queen. Another 10% chance is rolled for the queen objective if you get the trophy one instead.
- Round end results include who was specifically hunted down by the Yautja (crew wise). Predator objectives are already presented after the round ends.
- Neurotoxin Spit now respects shields.
- Aliens from different colonies can now attack each other.
- Aliens "wounded" chance lowered from 20 to 5. This was a chance rolled after an Alien attacked a human to stun them. Awful stuff.
- Queens can now toggle hivebalancing (forces you to evolve into the class with a fewer number of xenomorphs).
- Xenomorphs can tell where the Queen is by using their new alert (top right part of their screen).
- Firing ballistic guns within 4 turfs of a xenomorph will cause it to spray acidic blood back at you. Xenomorph blood has the potential of disintegrating your clothes, and burning you intensely.
- The priority announcement (attention: xenomorphs on board) only goes off after the xenomorphs have successfully captured and breeded a new larva.
- Rewrote end round conditions to support both the conquer and infestation objectives.
- Xenomorph Mode has a new objective: Infestation - infest 45/75% of the crew. Everytime someone dies (larva gives birth) from infestation, and is in the "target list", the hive will be notified of their success. Only people on the target list are used for the infestation objective.
- I might make a third alien objective, but it wouldn't involve any queens or workers. Just Hunters. And it'd be a lot less fun too. So I'll hold off on that.
- When the Queen leaves zlevel 1 (dead or alive), the shuttle will no longer be blocked and the crew can call it.
- Added a resin door which can only be opened up by members of the creators colony. In the event that a human (implanted organs) creates it, anyone can open the door.
- When created alien embryos won't ask ghosts to fill in potential larva positions. I personally think it'd be too much of a waiting game, and less entertaining.

What I'm still working on:
- Re-adding xeno spawns onto an updated map
- Updating z2 for the 3rd time w/ predator ship and droppod.
- Doublechecking the infest objective.

What to look out for:
- AFK Queen check and a braindead / logged out Queen check.
- Predators items/weapons are missing a LOT of action button sprites.
- I need help from a spriter with a few things KMC missed (non action button, mostly coloring).

ShadowDeath6 commented on code/modules/mob/living/carbon/alien/special/alien_embryo.dm:
```diff
@@ -34,34 +35,34 @@ var/const/ALIEN_AFK_BRACKET = 450 // 45 seconds
+	if(!embryo.key)
+		embryo.key = C.key
+		embryo << "<span class='alertalien'>Darkness surrounds you, and you grow bigger as you drain the nutrients out of your host. In time you'll soon be a fully grown...</span>"
+		embryo << "<span class='alertalien'>For now you are only a fetush slowly regenerating...</span>"
```
fetush?

ShadowDeath6 commented on code/modules/mob/living/carbon/alien/special/alien_embryo.dm:
```diff
@@ -111,23 +112,67 @@ var/const/ALIEN_AFK_BRACKET = 450 // 45 seconds
+/obj/item/organ/body_egg/alien_embryo/proc/join_infest_count()
+	if(ticker.mode.queensuffix == colony)
+		if(owner in ticker.mode.living_alien_targets)
+			ticker.mode.message_xenomorphs("Your hive has successfully transformed another targetted human!", FALSE, "alienannounce")
```
![target](https://cloud.githubusercontent.com/assets/17034893/26278857/ffdba4ae-3d69-11e7-9f2a-67515d4b5e7e.PNG)

ShadowDeath6 commented:
when you say that embryos don't ask for ghosts does that mean it forces people into it like the old system
if so don't do that

also would it be possible to add a "sense air" thing for ayys to know if an area is safe for hosts or not since their death will fuck the xenos over

I also think that you should make it so that if the person is suicided then the larva still bursts because fuck them breaking out for one second and pressing a macro that instakills them

Super3222 commented:
> when you say that embryos don't ask for ghosts does that mean it forces people into it like the old system

No, when I initially created this mode I made a system where the moment an embryo was created inside of a human it would poll ghosts to become the growing larva. The only problem is that it's a really unnecessary waiting game, so NOW (since the latest commit) it will only poll them when the embryo is about to burst.

> also would it be possible to add a "sense air" thing for ayys to know if an area is safe for hosts or not since their death will fuck the xenos over

What do you mean by "sense air"? As in a headcount for how many humans are in the area? If so that could be pretty interesting.

> I also think that you should make it so that if the person is suicided then the larva still bursts because fuck them breaking out for one second and pressing a macro that instakills them

True, I'll look for a check that tells whether someone killed themselves or not. There probably is one given that you can examine corpses and find out.

ShadowDeath6 commented:
sense air as in see how much oxygen/whatever is in the air so they can tell if it's habitable for humangs

X-TheDark commented on code/game/gamemodes/xenomorph/predatorgear/heartsensor.dm:
```diff
@@ -52,7 +52,10 @@
+	var/range_increment = 7
+	var/max_range = 21
+	for (var/i = start_range, i <= max_range, i += range_increment)
 		for(var/mob/living/carbon/human/H in orange(range, T))
```
You'd either need to rename `i` to `range` in line 58
OR
Rename `range` to `i` in line 59

Super3222 replied:
Ah, my bad. Fixing it in the next commit.

Super3222 commented:
A few notes to make:
- I'm weary about Predators. One one hand if done right they can be a pretty great asset to the mode, but on the other hand in a non-CM setting, they might be awful. Unsure about Goons handling on them, but from what I know they're a really rare antag that continuously hunts down the crew to earn points. My problem is that on here they might be a little too powerful and become a strong murderboning role. They might also be unfair for the normal crew to fight against - even with my nerfs and leeways to beat the Predators.
- Printing player doesn't work for whatever reason, so the round end information doesn't tell you who was the xenomorph queen. I'll look into this later.
- I could still use the help of a spriter to icon & color one or two things. It'll be easy to do.
- I'd appreciate code review unless you (Cruix and X) want to test this right ahead and see which bugs wash up on the surface.

Daman453 commented:
Why is this closed

Super3222 commented:
@Cruix Can you look at this? Your new PR just gave me another merge conflict.

Cruix commented:
Looks fine

Super3222 commented:
Some more just came in. I'll handle them when I can.

Super3222 commented:
- Fixed some xeno worker bugs
- Made adjustments to infestation objective
- Since we're not throwing ghosts into embryos before they burst, I removed the commented out code. I might later do something like CM did where ghosts can possess inactive larvas.
- Predators can now be summoned in the infestation objective. Much like the conquer one, if the xenos are 45% close to their goal the predators are called.
- Fixed xenomorph hotkey bug (merge conflict issue)
- I commented out the conquer objective only to test the infestation one. next commit that'll be fixed

Still have to fix the new merge conflicts. On the plus side, less pr's being made/merged nowadays = less stuff to fix in this pr.

Super3222 commented:
I've so far submitted 16 merge conflict commits. I don't have a lot of motivation to complete my 17th.

Jc1469 commented:
https://www.youtube.com/watch?v=KxGRhd_iWuE

Time-Green commented:
Happy 4 months anniversairy

Kmc2000 commented:
Happy anniversairy!

Super3222 commented:
I'm going to try to work with Cruix on how I should atomize this PR so we can get it done and merged faster.
